preview_generator/preview/builder/image__pillow.py

```diff
@@ -1,16 +1,193 @@
 from preview_generator.utils import ImgDims
 
+# INFO - G.M - 2019-06-25 - some configuration for jpeg saved,
+# see https://pillow.readthedocs.io/en/latest/handbook/image-file-formats.html?highlight=JPEG#jpeg
+# optimize -> extra pass to select optimal encoded
+# quality -> compression level from 1 to 100
+# progressive -> permit to see a blured version of file instead of partial image when image
+# is not fully downloaded, nice for web usage.
+DEFAULT_JPEG_OPTIMIZE = True
+DEFAULT_JPEG_QUALITY = 75
+DEFAULT_JPEG_PROGRESSIVE = True
+# Pillow algorithm use for resampling, bilinear is not the fastest but is fast
+# enough, this permit good enough image (using faster algorithm return much more
+# aliasing. see https://pillow.readthedocs.io/en/latest/handbook/concepts.html?highlight=Bilinear#filters
+# for algorithm list.
+DEFAULT_JPEG_RESAMPLE_ALGORITHM = PIL.Image.BILINEAR
+# Pillow deal with many different mode of image:
+# https://pillow.readthedocs.io/en/latest/handbook/concepts.html#modes
+# those are mode known to have alpha layer, so needing special process to
+# have clean white background
+# RGBA is RGB + Alpha channel
+# LA is L (8-bit pixels, black and white) + Alpha channel
+# color for
+DEFAULT_RGB_BACKGROUND_COLOR = (255, 255, 255) # This means white background
+DEFAULT_BW_BACKGROUND_COLOR = 1
+DEFAULT_SAVING_MODE = 'RGB'
+
+
+class ImageConvertStrategy(ABC):
+
+    @abstractmethod
+    def save(
+            self,
+            origin_image: PIL.Image,
+            file_output: BytesIO,
+            optimize: bool = DEFAULT_JPEG_OPTIMIZE,
+            quality: int = DEFAULT_JPEG_QUALITY,
+            progressive: bool= DEFAULT_JPEG_PROGRESSIVE
+    ) -> BytesIO:
+        pass
+
+class NotTransparentImageConvertStrategy(ImageConvertStrategy):
+    """
+    Image converter for all non transparent image
+    """
+
+    def save(
+            self,
+            origin_image: PIL.Image,
+            file_output: BytesIO,
+            optimize: bool = DEFAULT_JPEG_OPTIMIZE,
+            quality: int = DEFAULT_JPEG_QUALITY,
+            progressive: bool= DEFAULT_JPEG_PROGRESSIVE
+    ) -> BytesIO:
+        try:
+            origin_image.save(fp=file_output, format='jpeg', optimize=optimize, quality=quality,
+                              progressive=progressive)
+        except OSError as exc:
+            # INFO - G.M - in some case image mode cannot be directly convert to JPEG, in those
+            # case, it raise OSError, we should fallback to a working mode and retry saving.
+            origin_image = origin_image.convert(DEFAULT_SAVING_MODE)
+            origin_image.save(fp=file_output, format='jpeg', optimize=optimize, quality=quality,
+                              progressive=progressive)
+        file_output.seek(0, 0)
+        return file_output
+
+class TransparentImageConvertStrategy(ImageConvertStrategy):
+    """
+    Abstract image converter for Transparent Image
+    """
+
+    def save_transparency(
```

Comment:
this method is protected, no? it is not intended to be used from outside? if so, you should name it with `_` prefix

Reply:
better rename `save_transparency` to `save_with_transparency` or `save_transparent_image`

Reply:
done